File: keygen.py
#!/usr/bin/env python3
"""
keygen.py
---------
مولد توقيع ترخيص / تفعيل التطبيق وفق الشروط المحددة.

شروط الترخيص:
  • Endpoint:        POST /validate
  • Max URL length:  52 حرفاً
  • Response fields: valid, reason, expires, sig
  • Message format:  "<valid>|<reason>|<expires>"   (UTF-8)
  • Signature:       HMAC-SHA256, hex (64 char)
  • Secret (hex):    b309258d2e096b3bdfece839d7b38c936ae7b9af25ab775a70a358d9d921c05e
"""

import hmac
import hashlib
import json


# ----------------------------------------------------------------------
# ثوابت الترخيص
# ----------------------------------------------------------------------
SECRET_HEX = "b309258d2e096b3bdfece839d7b38c936ae7b9af25ab775a70a358d9d921c05e"
SECRET_KEY = bytes.fromhex(SECRET_HEX)          # تحويل السر إلى بايتات

ENDPOINT_PATH = "/validate"
MAX_URL_LEN   = 52                              # الحد الأقصى لطول الرابط الكامل


# ----------------------------------------------------------------------
# دوال التوليد
# ----------------------------------------------------------------------
def build_message(valid: bool, reason: str, expires: str) -> str:
    """
    يبني السلسلة الخام التي سيتم توقيعها.
    التنسيق:  "true|none|2028-12-31"
    """
    return f"{'true' if valid else 'false'}|{reason}|{expires}"


def sign_message(message: str, key: bytes = SECRET_KEY) -> str:
    """
    يوقّع الرسالة باستخدام HMAC-SHA256 ويُرجع التوقيع بصيغة Hex (64 حرفاً).
    """
    mac = hmac.new(key, message.encode("utf-8"), hashlib.sha256)
    return mac.hexdigest()


def generate_license(valid: bool = True,
                     reason: str = "none",
                     expires: str = "2028-12-31") -> dict:
    """
    يُولّد كائن JSON الكامل للرد على نقطة /validate.
    """
    message = build_message(valid, reason, expires)
    signature = sign_message(message)

    response = {
        "valid":   valid,
        "reason":  reason,
        "expires": expires,
        "sig":     signature,
    }
    return response


def verify_license(payload: dict) -> bool:
    """
    يتحقّق من صحة توقيع رد الترخيص.
    """
    expected_sig = sign_message(
        build_message(payload["valid"], payload["reason"], payload["expires"])
    )
    return hmac.compare_digest(expected_sig, payload.get("sig", ""))


# ----------------------------------------------------------------------
# نقطة التشغيل
# ----------------------------------------------------------------------
def main() -> None:
    license_obj = generate_license(
        valid=True,
        reason="ok",
        expires="2099-12-31T23:59:59Z",
    )

    raw_message = build_message(
        license_obj["valid"], license_obj["reason"], license_obj["expires"]
    )

    print("=" * 60)
    print("  License / Activation Key Generator")
    print("=" * 60)
    print(f"Endpoint        : POST {ENDPOINT_PATH}")
    print(f"Max URL length  : {MAX_URL_LEN} chars")
    print(f"Raw message     : {raw_message}")
    print(f"Signature (hex) : {license_obj['sig']}  (len={len(license_obj['sig'])})")
    print("-" * 60)
    print("JSON Response:")
    print(json.dumps(license_obj, ensure_ascii=False, indent=2))
    print("-" * 60)
    print(f"Verification    : {'VALID ✓' if verify_license(license_obj) else 'INVALID ✗'}")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: keygen.js
#!/usr/bin/env node
import crypto from "node:crypto";
import readline from "node:readline";
import fs from "node:fs/promises";
import { existsSync } from "node:fs";

const KEYS_FILE = "./keys.json";
const SECRET_HEX = "b309258d2e096b3bdfece839d7b38c936ae7b9af25ab775a70a358d9d921c05e";

// ═══════════════════════════════════════════════════════════
// Helper Functions
// ═══════════════════════════════════════════════════════════

function generateRandomKey() {
  const prefix = ["GOLD", "PREMIUM", "PRO", "ELITE"][Math.floor(Math.random() * 4)];
  const year = new Date().getFullYear();
  const random1 = crypto.randomBytes(4).toString("hex").toUpperCase();
  const random2 = crypto.randomBytes(4).toString("hex").toUpperCase();
  return `${prefix}-${year}-${random1}-${random2}`;
}

function generateRandomHWID() {
  return `HWID-${crypto.randomBytes(8).toString("hex").toUpperCase()}`;
}

async function loadKeys() {
  try {
    if (!existsSync(KEYS_FILE)) {
      return { keys: [] };
    }
    const data = await fs.readFile(KEYS_FILE, "utf8");
    return JSON.parse(data);
  } catch (error) {
    console.error("❌ Error loading keys:", error.message);
    return { keys: [] };
  }
}

async function saveKeys(keysData) {
  try {
    await fs.writeFile(KEYS_FILE, JSON.stringify(keysData, null, 2));
    return true;
  } catch (error) {
    console.error("❌ Error saving keys:", error.message);
    return false;
  }
}

function createSignature(message) {
  const keyBuffer = Buffer.from(SECRET_HEX, "hex");
  return crypto.createHmac("sha256", keyBuffer).update(message).digest("hex");
}

function buildResponse(valid, reason, expires) {
  const validStr = valid ? "true" : "false";
  const signedMessage = `${validStr}|${reason}|${expires}`;
  const sig = createSignature(signedMessage);

  return {
    valid,
    reason,
    expires,
    sig
  };
}

// ═══════════════════════════════════════════════════════════
// Input Helper
// ═══════════════════════════════════════════════════════════

function question(prompt) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  return new Promise((resolve) => {
    rl.question(prompt, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

// ═══════════════════════════════════════════════════════════
// Main Program
// ═══════════════════════════════════════════════════════════

async function main() {
  console.log("\n" + "═".repeat(70));
  console.log("  🔑 Golden Auth - Key Generator v2.0");
  console.log("═".repeat(70) + "\n");

  console.log("📝 Choose an option:");
  console.log("  1. Generate new key (interactive)");
  console.log("  2. Generate new key (automatic)");
  console.log("  3. List all keys");
  console.log("  4. Test signature generation");
  console.log("  5. Exit\n");

  const choice = await question("➡️  Enter choice (1-5): ");

  switch (choice) {
    case "1":
      await generateInteractive();
      break;
    case "2":
      await generateAutomatic();
      break;
    case "3":
      await listKeys();
      break;
    case "4":
      await testSignature();
      break;
    case "5":
      console.log("\n👋 Goodbye!\n");
      process.exit(0);
    default:
      console.log("❌ Invalid choice");
      process.exit(1);
  }
}

async function generateInteractive() {
  console.log("\n🎯 Interactive Key Generation\n");

  const key = await question(`Key (Enter for random): `);
  const hwid = await question(`HWID (Enter for random): `);
  const expiresInput = await question(`Expires (YYYY-MM-DD or 'never'): `);
  const note = await question(`Note (optional): `);

  const finalKey = key || generateRandomKey();
  const finalHWID = hwid || generateRandomHWID();
  
  let finalExpires;
  if (expiresInput.toLowerCase() === "never" || !expiresInput) {
    finalExpires = "2099-12-31T23:59:59Z";
  } else {
    finalExpires = `${expiresInput}T23:59:59Z`;
  }

  const newKeyEntry = {
    key: finalKey,
    hwid: finalHWID,
    expires: finalExpires,
    active: true,
    created: new Date().toISOString(),
    uses: 0,
    note: note || "Generated via keygen"
  };

  const keysData = await loadKeys();
  keysData.keys.push(newKeyEntry);
  
  if (await saveKeys(keysData)) {
    console.log("\n" + "═".repeat(70));
    console.log("✅ Key Generated Successfully!");
    console.log("═".repeat(70));
    console.log(`Key:     ${newKeyEntry.key}`);
    console.log(`HWID:    ${newKeyEntry.hwid}`);
    console.log(`Expires: ${newKeyEntry.expires}`);
    console.log(`Note:    ${newKeyEntry.note}`);
    console.log("═".repeat(70) + "\n");
  }
}

async function generateAutomatic() {
  console.log("\n⚡ Automatic Key Generation\n");

  const count = await question(`How many keys to generate? (1-100): `);
  const numKeys = Math.min(100, Math.max(1, parseInt(count) || 1));

  const keysData = await loadKeys();
  const newKeys = [];

  for (let i = 0; i < numKeys; i++) {
    const newKeyEntry = {
      key: generateRandomKey(),
      hwid: generateRandomHWID(),
      expires: "2099-12-31T23:59:59Z",
      active: true,
      created: new Date().toISOString(),
      uses: 0,
      note: `Auto-generated ${i + 1}/${numKeys}`
    };
    keysData.keys.push(newKeyEntry);
    newKeys.push(newKeyEntry);
  }

  if (await saveKeys(keysData)) {
    console.log(`\n✅ Generated ${numKeys} keys successfully!\n`);
    newKeys.forEach((k, i) => {
      console.log(`${i + 1}. ${k.key} | ${k.hwid}`);
    });
    console.log("");
  }
}

async function listKeys() {
  console.log("\n📊 All Keys\n");

  const keysData = await loadKeys();
  
  if (keysData.keys.length === 0) {
    console.log("❌ No keys found\n");
    return;
  }

  console.log("═".repeat(100));
  console.log(`Total: ${keysData.keys.length} keys\n`);
  
  keysData.keys.forEach((k, i) => {
    console.log(`[${i + 1}] ${k.active ? '✅' : '❌'} ${k.key}`);
    console.log(`    HWID:    ${k.hwid}`);
    console.log(`    Expires: ${k.expires}`);
    console.log(`    Uses:    ${k.uses || 0}`);
    console.log(`    Note:    ${k.note || 'N/A'}`);
    console.log(``);
  });
  console.log("═".repeat(100) + "\n");
}

async function testSignature() {
  console.log("\n🔬 Test Signature Generation\n");

  const valid = true;
  const reason = "ok";
  const expires = "2099-12-31T23:59:59Z";

  const response = buildResponse(valid, reason, expires);

  console.log("═".repeat(70));
  console.log("Test License Response:");
  console.log("═".repeat(70));
  console.log(JSON.stringify(response, null, 2));
  console.log("═".repeat(70));
  console.log(`\nSignature length: ${response.sig.length} chars`);
  console.log(`Expected: 64 chars (HMAC-SHA256 hex)\n`);
}

// Run
main().catch(console.error);
